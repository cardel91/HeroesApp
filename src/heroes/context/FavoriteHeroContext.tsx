import { createContext, useEffect, useState, type PropsWithChildren } from "react"
import type { Hero } from "../types/hero.interface";

interface FavoriteHeroContext {
    favorites: Hero[];
    favoriteCount: number;

    isFavorite: (hero: Hero) => boolean;
    toggleFavorite: (hero: Hero) => void;


}

// eslint-disable-next-line react-refresh/only-export-components
export const FavoriteHeroContext = createContext({} as FavoriteHeroContext);

const getFavoritesFromLocal = (): Hero[] => {
    const favorites = localStorage.getItem('favorites');
    return favorites ? JSON.parse(favorites) : [];
}


export const FavoriteHeroProvider = ({ children }: PropsWithChildren) => {

    const [favorites, setFavorites] = useState<Hero[]>(
        getFavoritesFromLocal()
    );

    const toggleFavorite = (hero: Hero) => {
        const heroExists = favorites.find((h) => h.id === hero.id);


        if (heroExists) {
            const newFavorites = favorites.filter((h) => h.id !== hero.id);
            setFavorites(newFavorites);
            return;
        }

        setFavorites([...favorites, hero]);
    }

    const isFavorite = (hero: Hero) => favorites.some((h) => h.id === hero.id)

    useEffect(() => {
        localStorage.setItem('favorites', JSON.stringify(favorites));
    }, [favorites])


    return (
        <FavoriteHeroContext
            value={{
                favoriteCount: favorites.length,
                favorites: favorites,
                isFavorite: isFavorite,
                toggleFavorite: toggleFavorite
            }}
        >
            {children}
        </FavoriteHeroContext>
    )
}


