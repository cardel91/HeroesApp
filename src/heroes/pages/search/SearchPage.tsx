import { HeadComponent } from "@/components/custom/HeadComponent";
import { HeroStats } from "@/heroes/components/HeroStats";
import { SearchControls } from "./ui/SearchControls";


const SearchPage = () => {
    return (
        <>
            <HeadComponent title="Universo de heroes" description="Descubre, explota y descarta super heroes y villanos" />
            <HeroStats />

            <SearchControls />
        </>
    )
}

export default SearchPage;
