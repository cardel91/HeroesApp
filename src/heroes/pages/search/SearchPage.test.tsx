import { FavoriteHeroProvider } from "@/heroes/context/FavoriteHeroContext";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import { beforeEach, describe, expect, test, vi } from "vitest";
import SearchPage from "./SearchPage";
import { searchHeroesAction } from "@/heroes/actions/search-heroes.action";
import type { Hero } from "@/heroes/types/hero.interface";

const queryClient = new QueryClient();

const renderSearch = (initialEntries: string[] = ['/']) => {

    return render(
        <MemoryRouter initialEntries={initialEntries}>
            <FavoriteHeroProvider>
                <QueryClientProvider client={queryClient}>
                    <SearchPage />
                </QueryClientProvider>
            </FavoriteHeroProvider>
        </MemoryRouter>
    )
}

vi.mock('@/heroes/actions/search-heroes.action');

const searchHeroesActionMock = vi.mocked(searchHeroesAction);

vi.mock('@/components/custom/HeadComponent', () => ({
    HeadComponent: () => <div data-testid="custom-head"></div>
}));


vi.mock('./ui/SearchControls', () => ({
    SearchControls: () => <div data-testid="search-controls"></div>
}))


vi.mock('@/heroes/components/HeroGrid', () => ({
    HeroGrid: ({ heroes }: { heroes: Hero[] }) => (<div data-testid="custom-hero-grid">
        {
            heroes.map((hero) => (
                <div key={hero.id}>{hero.name}</div>
            ))
        }
    </div>)
}));

describe('SearchPage', () => {


    beforeEach(() => {
        vi.clearAllMocks();
    });

    test('should render SearchPage with default values', () => {

        const { container } = renderSearch();

        expect(searchHeroesActionMock).toHaveBeenCalledWith({
            name: undefined,
            strength: 0
        });

        expect(container).toMatchSnapshot();

    });

    test('should call search action with name parameter', () => {

        const { container } = renderSearch(['/search?name=superman']);

        expect(searchHeroesActionMock).toHaveBeenCalledWith({
            name: 'superman',
            strength: 0
        });

        expect(container).toMatchSnapshot();

    });

    test('should call search action with strength parameter', () => {

        const { container } = renderSearch(['/search?strength=5']);

        expect(searchHeroesActionMock).toHaveBeenCalledWith({
            name: undefined,
            strength: 5
        });

        expect(container).toMatchSnapshot();

    });

    test('should call search action with name and strength parameters', () => {

        const { container } = renderSearch(['/search?strength=5&name=Joker']);

        expect(searchHeroesActionMock).toHaveBeenCalledWith({
            name: 'Joker',
            strength: 5
        });

        expect(container).toMatchSnapshot();

    });

    test('should render HeroGrid with search results', async () => {
        const mockedHeroes = [
            { id: '1', name: "Clark Kenton" } as unknown as Hero,
            { id: '2', name: "Bruce Waylon" } as unknown as Hero,
        ];

        searchHeroesActionMock.mockResolvedValue(mockedHeroes);
        renderSearch();

        await waitFor(() => {
            expect(screen.getByText('Clark Kenton')).toBeDefined();
            expect(screen.getByText('Bruce Waylon')).toBeDefined();
        });

        screen.debug(screen.getByTestId('custom-hero-grid'));
    });

});