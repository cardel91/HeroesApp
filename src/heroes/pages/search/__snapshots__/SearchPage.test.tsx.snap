// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchPage > should call search action with name and strength parameters 1`] = `
<div>
  <div
    data-testid="custom-head"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/search"
        >
          Search
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes/1"
        >
          Hero
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="text-foreground font-normal"
          data-slot="breadcrumb-page"
          role="link"
        >
          Search
        </span>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="custom-hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with name parameter 1`] = `
<div>
  <div
    data-testid="custom-head"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/search"
        >
          Search
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes/1"
        >
          Hero
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="text-foreground font-normal"
          data-slot="breadcrumb-page"
          role="link"
        >
          Search
        </span>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="custom-hero-grid"
  />
</div>
`;

exports[`SearchPage > should call search action with strength parameter 1`] = `
<div>
  <div
    data-testid="custom-head"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/search"
        >
          Search
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes/1"
        >
          Hero
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="text-foreground font-normal"
          data-slot="breadcrumb-page"
          role="link"
        >
          Search
        </span>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="custom-hero-grid"
  />
</div>
`;

exports[`SearchPage > should render SearchPage with default values 1`] = `
<div>
  <div
    data-testid="custom-head"
  />
  <nav
    aria-label="breadcrumb"
    class="my-5"
    data-slot="breadcrumb"
  >
    <ol
      class="text-muted-foreground flex flex-wrap items-center gap-1.5 text-sm break-words sm:gap-2.5"
      data-slot="breadcrumb-list"
    >
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/search"
        >
          Search
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <a
          class="hover:text-foreground transition-colors"
          data-discover="true"
          data-slot="breadcrumb-link"
          href="/heroes/1"
        >
          Hero
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:size-3.5"
        data-slot="breadcrumb-separator"
        role="presentation"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-chevron-right"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m9 18 6-6-6-6"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
        data-slot="breadcrumb-item"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="text-foreground font-normal"
          data-slot="breadcrumb-page"
          role="link"
        >
          Search
        </span>
      </li>
    </ol>
  </nav>
  <div>
    Loading...
  </div>
  <div
    data-testid="search-controls"
  />
  <div
    data-testid="custom-hero-grid"
  />
</div>
`;
